### PR TITLE
Support security encoder and added support for argon2 and co.

### PR DESCRIPTION
**Pull request by Toflar**
In Symfony 4.3, a new security encoder named `auto` was introduced. It will choose the best password hashing algorithm based on what Symfony and/or the PHP community decides.

Currently this would mean as of PHP 7.2 if you did compile your PHP using libsodium, you'll have argon2 available. I wanted to make sure my setup benefits from the strongest algorithms available so I've implemented the `auto` configuration by default and it will fall back to the current `bcrypt` configuration in case you are not yet on Symfony 4.3. Of course `auto` may still take `bcrypt` in case your system does not support libsodium.

While implementing I noticed that we still use the password hashing API directly instead of relying on the Symfony components. This means you can configure a different algorithm in the security configuration but it will never work 😄 So I fixed this by replacing all our direct password hashing API calls with the underlying Symfony security components.

BTW: I had to remove the usage of `password_needs_rehash()` to completely rely on the Symfony encoder configured. This means right now you cannot automatically upgrade which is why currently the `SodiumPasswordEncoder` will also verify bcrypt passwords. Symfony will address this in 4.4 so we can (https://github.com/symfony/symfony/pull/31843) completely rely on what Symfony is doing very soon. As of now, there are only `bcrypt` and `argon2` anyway and those will be handled properly.

### CODE REVIEW

File: core-bundle/src/Resources/contao/forms/FormPassword.php [hunk 1]
```diff
@@ -139,7 +140,11 @@ protected function validator($varInput)
 		{
 			$this->blnSubmitInput = true;
 
-			return password_hash($varInput, PASSWORD_DEFAULT);
+			/** @var EncoderFactoryInterface $encoderFactory */
+			$encoderFactory = System::getContainer()->get('security.encoder_factory');
+			$encoder = $encoderFactory->getEncoder(BackendUser::class);
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/forms/FormPassword.php]: This should probably be `FrontendUser` or just `Contao\User`

File: core-bundle/src/Resources/contao/library/Contao/Encryption.php [hunk 2]
```diff
@@ -168,7 +170,11 @@ protected static function initialize()
 	 */
 	public static function hash($strPassword)
 	{
-		return password_hash($strPassword, PASSWORD_DEFAULT);
+		/** @var EncoderFactoryInterface $encoderFactory */
+		$encoderFactory = System::getContainer()->get('security.encoder_factory');
+		$encoder = $encoderFactory->getEncoder(BackendUser::class); // This is hardcoded (BE and FE users/members should not have a different algo anyway but if you need it, do not rely on a deprecated class but use the services instead)
```
Comment (aschempp) [inline core-bundle/src/Resources/contao/library/Contao/Encryption.php]: Should use `Contao\User` imho

**Review by aschempp — APPROVED** [review]
LGTM except for the class usages

**Review by leofeyer — CHANGES_REQUESTED** [review]
First of all, this change implies dropping support for Symfony 4.2, so the `composer.json` file needs to be adjusted.

> BTW: I had to remove the usage of `password_needs_rehash()`

So how to old password hashes get updated and persisted now?

**Comment by Toflar** [comment]
> First of all, this change implies dropping support for Symfony 4.2, so the `composer.json` file needs to be adjusted.

That's not true. It continues using `bcrypt` if you have < 4.3.



> So how to old password hashes get updated and persisted now?

They don't. We can only implement that when Symfony 4.4 is released. But we're talking about `bcrypt` to something else and "something else" does not exist in PHP yet. The `PASSWORD_DEFAULT` constant is still configured to bcrypt so nothing had to be upgraded so far. This PR would only mean a regression if PHP updated their constant which they did not yet and there are no plans for 7.4 either. So by the time they'll do that we'll have Symfony 4.4 and changed already :)

File: core-bundle/src/Resources/contao/controllers/BackendPassword.php [hunk 3]
```diff
@@ -81,8 +82,12 @@ public function run()
 			// Save the data
 			else
 			{
+				/** @var EncoderFactoryInterface $encoderFactory */
+				$encoderFactory = System::getContainer()->get('security.encoder_factory');
+				$encoder = $encoderFactory->getEncoder(BackendUser::class);
+
 				// Make sure the password has been changed
-				if (password_verify($pw, $this->User->password))
+				if ($encoder->isPasswordValid($this->User->password, $pw, null))
```
Comment (leofeyer) [inline core-bundle/src/Resources/contao/controllers/BackendPassword.php:87]: Why did you change the argument order?

Comment (Toflar) [reply core-bundle/src/Resources/contao/controllers/BackendPassword.php:87]: Because that's what the interface expects.

**Comment by leofeyer** [comment]
That's clever 🦊 

Then the only thing that we need to discuss is the class usage. There are still a lot of `getEncoder(BackendUser::class)` and `getEncoder(FrontendUser::class)` calls, although both use the same algorithm and we could use `getEncoder(User::class)` everywhere.

**Comment by Toflar** [comment]
Yeah in fact it doesn't really matter. It's just "more correct" the way it is now 😄

**Comment by leofeyer** [comment]
Note to myself: After merging this, create a ticket for Contao 4.9 reminding us to implement the `PasswordUpgraderInterface` interface (see https://github.com/symfony/symfony/pull/31843).

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @Toflar.

**Comment by leofeyer** [comment]
Reminder issue: #538

**Comment by richardhj** [comment]
https://github.com/contao/contao/blob/f5748dbb0a0a891ee12fc63656c0d7ed6a4dcd8b/core-bundle/src/Resources/contao/library/Contao/Encryption.php#L183-L185
`Encryption::test()` provides a method to test if a given string is a hash.

Recently I've been using `if (strncmp($strHash, '$2y$', 4) === 0)` to check if a string is already hashed. 

1. Should `Encryption::test()` be updated, even though it is deprecated?
2. How do I check if a string is already hashed?

**Comment by Toflar** [comment]
>     2\. How do I check if a string is already hashed?

Why do you need that anyway? :)

**Comment by richardhj** [comment]
> Why do you need that anyway? :)

In the `save_callback` on the password field (tl_member/tl_user) it might be the case that the omitted password is unhashed. If so, I need to hash the password. For that, I need to check if the password is already hashed :)

**Comment by Toflar** [comment]
I don't understand that use case, sorry :) But maybe you could do this in PHP when using the password hashing API using `password_get_info()`. I don't think you can in Symfony though.

**Comment by richardhj** [comment]
`password_get_info()` is totally fine! ☺️

**Comment by Toflar** [comment]
Yeah so there's no way to do so as of today. Time for a SF PR if you need that ;)

**Comment by ausi** [comment]
> How do I check if a string is already hashed?

Doing this is probably be a security issue because a password that looks like a hash must not be handled differently. What if my actual password starts with `$2y$`?

I think you have to design your code so that you know in advance if you are working with a hash or with a plaintext password.
